lib/Parser/BGParseManager.cpp

```diff
@@ -0,0 +1,467 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#include "ParserPch.h"
+
+#include "Memory/AutoPtr.h"
+#include "Common/Event.h"
+#include "Base/ThreadContextTlsEntry.h"
+#include "base/ThreadBoundThreadContextManager.h"
+#include "Base/Utf8SourceInfo.h"
+#include "BGParseManager.h"
+#include "Base/ScriptContext.h"
+#include "ByteCodeSerializer.h"
+
+// Global, process singleton
+BGParseManager* BGParseManager::s_BGParseManager = nullptr;
+DWORD           BGParseManager::s_lastCookie = 0;
+CriticalSection BGParseManager::s_staticMemberLock;
+
+// Static member management
+BGParseManager* BGParseManager::GetBGParseManager()
+{
+    AutoCriticalSection lock(&s_staticMemberLock);
+    if (s_BGParseManager == nullptr)
+    {        
+        AUTO_NESTED_HANDLED_EXCEPTION_TYPE(ExceptionType_DisableCheck);
+        s_BGParseManager = HeapNew(BGParseManager);
+        s_BGParseManager->Processor()->AddManager(s_BGParseManager);
+    }
+    return s_BGParseManager;
+}
+
+void BGParseManager::DeleteBGParseManager()
+{
+    AutoCriticalSection lock(&s_staticMemberLock);
+    if (s_BGParseManager != nullptr)
+    {
+        BGParseManager* manager = s_BGParseManager;
+        s_BGParseManager = nullptr;
+        HeapDelete(manager);
+    }
+}
+
+DWORD BGParseManager::GetNextCookie()
+{
+    AutoCriticalSection lock(&s_staticMemberLock);
+    return ++s_lastCookie;
+}
+
+
+// Note: runs on any thread
+BGParseManager::BGParseManager()
+    : JsUtil::WaitableJobManager(ThreadBoundThreadContextManager::GetSharedJobProcessor())
+{
+}
+
+// Note: runs on any thread
+BGParseManager::~BGParseManager()
+{
+    // First, remove the manager from the JobProcessor so that any remaining jobs
+    // are moved back into this manager's processed workitem list
+    if (Processor()->HasManager(this))
```

Comment:
>HasManager [](start = 21, length = 10)

The implementation of HasManager doesn't look thread-safe; is that a concern? In general I'd prefer updating RemoveManager so it does nothing if the manager is not present, rather than requiring callers to guarantee that the manager is present before requesting its removal.

Reply:
it wasn't a concern during the initial prototype, but I can't prove that it's not a concern now, so I assume that it will be a problem.
makes sense to update RemoveManager